### ISSUE TITLE
Use macros instead of colorMap for KA specific colors

### ISSUE THREAD

**Issue by kevinbarabash** [Khan/KaTeX#750]
See TODO in Options.js

**Comment by edemaine** [comment]
These would most naturally be replaced by `\definecolor` calls (see [this description](https://en.wikibooks.org/wiki/LaTeX/Colors#Defining_new_colors)).  Unless you want to replace `\textcolor{katex-blue}{...}` with `\textcolor{\katexblue}{...}` and use a macro `\katexblue` (but sadly can't have a hyphen in its name).  By analogy to the `macros` option we currently have, we could have a `colors` option that provides a mapping of colors to RGB.  This map could then be moved to the KA project where KaTeX gets called, or it could be the default settings which uses can add additional colors to.

**Comment by kevinbarabash** [comment]
Cool!  I didn't know about `\definecolor`.  I think implementing that to support at least the RGB and HTML models makes sense since we already have a `colorMap` object.

**Comment by danielhollas** [comment]
I wanted to give this one a try, although I am not sure this a good first issue :-)
My idea was to create a customColorMap object that would look like this:
```js
let customColorMap = {
  "myColor": "#ffaaff"
}
```

and modify `parseColorGroup` in Parser.js to take into account this object:
https://github.com/KaTeX/KaTeX/blob/b9deec098f35919b9ad77baa6b11b7593473d488/src/Parser.js#L649
```
let inputColor = res.text;
if (colorMap.hasOwnProperty(inputColor)) {
     inputColor = colorMap[inputColor];
}
```
So far so good and this works, but now I don't know how to properly define a function whose sole purpose would be to extend the customColorMap object and nothing else (no htmlbuilder etc). Is there a precedent here that I could follow?

**Comment by edemaine** [comment]
@danielhollas I believe this map would most naturally be an addition to the (recent) `Namespace` class/js file.  Looking at the definition of `\definecolor`, it looks like `\definecolor{red}{rgb}{1,0,0}` defines `\color@red` by calling `\color@rgb{\string\color@red}{1,0,0}`.  We could just simulate that behavior, or at least use the macro mapping by setting `\color@red` to the color.  Alternatively, you could try adding a special mapping for colors in the Namespace, though then you'll need to do more work when namespaces get pushed or popped.

Could you test whether `{\definecolor{myred}{rgb}{1,0,0}}\color{myred}` fails as I expect it should, i.e., the definition of color is local to the group?

**Comment by danielhollas** [comment]
Indeed, definecolor should be local to the group, I tested your command and it is also said in the documentation: http://mirrors.nic.cz/tex-archive/macros/latex/contrib/xcolor/xcolor.pdf

That said, my guess is most people would use this command in the preamble anyway. Also, since KaTeX processes math chunks individually, I don't see how `definecolor{}` would be of any use if you had to redefine it in each `$..$` element. Would you agree?

Thanks for pointing me to Namespace. It seems to allow global definitions as well. Still not sure where to put the `definecolor` function though so that it modifies this namespace. Perhaps this is a bit too complicated for me at this point.

**Comment by edemaine** [comment]
Hmm, good point; with `\def` we have `\gdef` to affect global scope (and thus affect future calls to KaTeX, using the `macros` object). For `\definecolor`, it looks like (according to the `xcolor` documentation you linked) LaTeX users are supposed to use `\xglobal\definecolor` or set `\globalcolorstrue` to make global color definitions.

We could add a `katex.defineColor` or similar method that does global color definitions from JavaScript. Or maybe better would be to add a `katex.globalEval` that runs some definitions at the level of the global namespace, which would also allow global `\newcommand` macros.  (Of course, that is probably beyond the scope of a `\definecolor` PR, but it's a future direction where this makes sense.)

**Comment by kevinbarabash** [comment]
I think having a common way to run a series of commands and have their effects be global is probably the way to go that way we don't have to add new methods to `katex` each time we want to do something global so my vote would be for adding `globalEval`.  More discussion about what exactly this function would do is probably warranted.

**Comment by Artoria2e5** [comment]
A related request is about the pre-defined colors and invoking them via `\definecolors` -- a colormap might be a natural fit for it. On Wikimedia the default is to always load dvipsnames, so some way to load an entire set is also required (since we can't quite do usepackage [can we?]).
